cpp/src/arrow/array/builder_base.cc

```diff
@@ -111,4 +115,43 @@ void ArrayBuilder::UnsafeSetNull(int64_t length) {
   null_bitmap_builder_.UnsafeAppend(length, false);
 }
 
+struct ArrayBuilderAppendScalarImpl {
+  template <typename S,
+            typename B = typename TypeTraits<typename S::TypeClass>::BuilderType>
+  B& builder() {
+    return *internal::checked_cast<B*>(builder_);
+  }
+
+  template <typename S>
+  auto Visit(const S& scalar) -> decltype(builder<S>().Append(scalar.value)) {
+    return builder<S>().Append(scalar.value);
+  }
+
+  template <typename S>
+  enable_if_base_binary<typename S::TypeClass, Status> Visit(const S& scalar) {
+    return builder<S>().Append(scalar.value->data(), scalar.value->size());
+  }
+
+  Status Visit(const Scalar& scalar) {
+    return Status::NotImplemented("Appending scalars to builders of type ", *scalar.type);
+  }
+
+  Status Finish() && { return VisitScalarInline(scalar_, this); }
+  ArrayBuilder* builder_;
+  const Scalar& scalar_;
+};
```

Comment:
If this is just to enable the test harness created below I'm not sure it's worth it